### ISSUE TITLE
Error when encode and decode by DALI

### ISSUE THREAD

**Issue by delamoerxy** [NVIDIA/DALI#256]
I'm confused that when I use tensorflow api to encode image,it can be decode by DALI correctly.
` 

     with tf.gfile.FastGFile(file_path, 'rb') as f
            image_raw = f.read()

        example = tf.train.Example(features=tf.train.Features(feature={
            'image_raw': tf.train.Feature(bytes_list=tf.train.BytesList(value=[image_raw])),
            'label': tf.train.Feature(int64_list=tf.train.Int64List(value=[label]))
        }))

        serialized = example.SerializeToString()
        tf_writer.write(serialized)`
But when I directly convert an 3D array tostring(), it can't be decode by DALI.
`        
       
        img = cv2.imread(filename)

        rgb_image = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

        image_raw = rgb_image.tostring()

        example = tf.train.Example(features=tf.train.Features(feature={
            'image_raw': _bytes_feature(image_raw),
            'label': _int64_feature(label)
        }))
        serialized =example.SerializeToString()
        writer.write(serialized)`
The error message is that

`
      nternalError (see above for traceback): DALI daliOutput(&pipe_handle_) failed: Critical error in pipeline: [/opt/dali/dali/pipeline/operators/decoder/nvjpeg_decoder.h:175] NVJPEG error "3"

Current pipeline object is no longer valid.`

**Comment by JanuszL** [comment]
Hi,
In the way you described, you are trying to use nvJPEG decoder to decode raw data. In the such case it is not able to determine image format as RAW data doesn't have any header (even if decoding RAW data have any sense).
Currently TFRecordReader won't work well either as it will return you just 1D tensor for you image instead of HWC one (one with defined width, height and number of channels). I think is it a good idea to extend it for such use case.
Tracked as DALI-320.

**Comment by JanuszL** [comment]
https://github.com/NVIDIA/DALI/pull/388 should help you-  after reading raw image you can reshape this 1D tensor to 2D image of given dimensions and skip decoding.

**Comment by JanuszL** [comment]
Also, you could define the shape of raw data in the TFRecord - tfrec.FixedLenFeature([X, Y], tfrec.int8,  -1). If shape changes from sample to sample then it would be difficult.

**Comment by JanuszL** [comment]
With the most recent DALI releases you can define image shape as a separate TFRecord fields, and then using [the cat` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.cat) merge them and then call [the `reshape` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.reshape) to add a desired dimensionality to the data.

**Comment by hexiaoyi95** [comment]
@JanuszL The reshape operator seems to only work for uint8 array (image) and cannot assign a "data type" argument. Do we have an operator like "tf.io.decode_raw" so we can decode raw data of  any types of array?

**Comment by JanuszL** [comment]
Hi @hexiaoyi95,

If you want to change the underlying data type please use [the `reinterpret` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.reinterpret).
